File: Exersice 5/Python/text_processor.py

from typing import Callable, List


def trim_text(text: str) -> str:
    return text.strip()

def to_upper(text: str) -> str:
    return text.upper()

def censor(text: str) -> str:
    forbidden = ["badword", "ugly"]
    for word in forbidden:
        text = text.replace(word, "***")
    return text

def prefix(text: str, pre: str) -> str:
    return pre + text


def part2():
    text = "   hello badword   "
    operations = [trim_text, to_upper, censor]
    for op in operations:
        print(op(text))


def process_text(funcs: List[Callable[[str], str]], text: str) -> str:
    for f in funcs:
        text = f(text)
    return text


def create_censor(forbidden_words: List[str]):
    def inner(text: str) -> str:
        for word in forbidden_words:
            text = text.replace(word, "***")
        return text
    return inner

def create_prefixer(pre: str):
    def inner(text: str) -> str:
        return pre + text
    return inner


def scope():
    x = "outer"
    def inner():
        nonlocal x
        x = "changed inside closure"
        print("Inner sees:", x)
    inner()
    print("Outer sees:", x)


def pipeline():
    texts = ["  badword detected  ", " ugly code  "]
    censor_custom = create_censor(["badword", "ugly"])
    prefixer = create_prefixer("NOTICE: ")

    pipeline = [trim_text, censor_custom, to_upper, prefixer]
    for t in texts:
        result = process_text(pipeline, t)
        print(f'Original: "{t}" → Processed: "{result}"')

if __name__ == "__main__":
    part2()
    scope()
    pipeline()

File: Exersice 5/Typescript/text_processor.ts

function trimText(text: string): string {
  return text.trim();
}

function toUpper(text: string): string {
  return text.toUpperCase();
}

function censor(text: string): string {
  const forbidden = ["badword", "ugly"];
  forbidden.forEach(w => text = text.replaceAll(w, "***"));
  return text;
}

function prefix(text: string, pre: string): string {
  return pre + text;
}


function part2() {
  const text = "   hello badword   ";
  const operations = [trimText, toUpper, censor];
  operations.forEach(op => console.log(op(text)));
}


function processText(funcs: ((t: string) => string)[], text: string): string {
  return funcs.reduce((acc, f) => f(acc), text);
}


function createCensor(forbidden: string[]): (text: string) => string {
  return (text: string) => {
    forbidden.forEach(w => text = text.replaceAll(w, "***"));
    return text;
  };
}

function createPrefixer(pre: string): (text: string) => string {
  return (text: string) => pre + text;
}


function scope() {
  let outer = "outer";
  function inner() {
    outer = "changed inside closure";
    console.log("Inner sees:", outer);
  }
  inner();
  console.log("Outer now:", outer);
}


function pipeline() {
  const texts = ["  badword detected  ", " ugly code  "];
  const censorCustom = createCensor(["badword", "ugly"]);
  const prefixer = createPrefixer("NOTICE: ");
  const pipeline = [trimText, censorCustom, toUpper, prefixer];

  texts.forEach(t => {
    const result = processText(pipeline, t);
    console.log(`Original: "${t}" → Processed: "${result}"`);
  });
}


part2();
scope();
pipeline();
